File: api/src/main/java/com/damier/damierclub/model/Article.java
package com.damier.damierclub.model;

import jakarta.persistence.*;
import org.hibernate.annotations.GenericGenerator;

import java.time.LocalDateTime;
import java.util.List;

@Entity
@Table(name = "articles")
public class Article {

    @Id
    @GeneratedValue(generator = "UUID")
    @GenericGenerator(name = "UUID", strategy = "org.hibernate.id.UUIDGenerator")
    @Column(name = "id", updatable = false, nullable = false, length = 36)
    private String id;

    @Column(nullable = false, length = 500)
    private String title;

    @Column(nullable = false, unique = true, length = 500)
    private String slug;

    @Column(columnDefinition = "TEXT", nullable = false)
    private String content; // JSON content from Editor.js

    @Column(columnDefinition = "TEXT")
    private String excerpt; // Short summary/description

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private ArticleCategory category;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private ArticleStatus status;

    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "author_id", nullable = false)
    private Member author;

    @Column(length = 1000)
    private String coverImage; // URL or path to cover image

    @ElementCollection
    @CollectionTable(name = "article_tags", joinColumns = @JoinColumn(name = "article_id"))
    @Column(name = "tag")
    private List<String> tags;

    @Column
    private LocalDateTime publishedAt;

    @Column(nullable = false)
    private LocalDateTime createdAt;

    @Column(nullable = false)
    private LocalDateTime updatedAt;

    @Column(nullable = false)
    private Integer viewCount = 0;

    @Column(nullable = false)
    private Boolean featured = false;

    // Enums
    public enum ArticleCategory {
        NEWS,
        RESULTS,
        EVENTS,
        TUTORIAL,
        ANNOUNCEMENT
    }

    public enum ArticleStatus {
        DRAFT,
        PUBLISHED,
        ARCHIVED
    }

    // Constructor
    public Article() {
        this.createdAt = LocalDateTime.now();
        this.updatedAt = LocalDateTime.now();
        this.viewCount = 0;
        this.featured = false;
        this.status = ArticleStatus.DRAFT;
    }

    // Lifecycle callbacks
    @PreUpdate
    protected void onUpdate() {
        this.updatedAt = LocalDateTime.now();
    }

    // Getters and Setters
    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public String getSlug() {
        return slug;
    }

    public void setSlug(String slug) {
        this.slug = slug;
    }

    public String getContent() {
        return content;
    }

    public void setContent(String content) {
        this.content = content;
    }

    public String getExcerpt() {
        return excerpt;
    }

    public void setExcerpt(String excerpt) {
        this.excerpt = excerpt;
    }

    public ArticleCategory getCategory() {
        return category;
    }

    public void setCategory(ArticleCategory category) {
        this.category = category;
    }

    public ArticleStatus getStatus() {
        return status;
    }

    public void setStatus(ArticleStatus status) {
        this.status = status;
    }

    public Member getAuthor() {
        return author;
    }

    public void setAuthor(Member author) {
        this.author = author;
    }

    public String getCoverImage() {
        return coverImage;
    }

    public void setCoverImage(String coverImage) {
        this.coverImage = coverImage;
    }

    public List<String> getTags() {
        return tags;
    }

    public void setTags(List<String> tags) {
        this.tags = tags;
    }

    public LocalDateTime getPublishedAt() {
        return publishedAt;
    }

    public void setPublishedAt(LocalDateTime publishedAt) {
        this.publishedAt = publishedAt;
    }

    public LocalDateTime getCreatedAt() {
        return createdAt;
    }

    public void setCreatedAt(LocalDateTime createdAt) {
        this.createdAt = createdAt;
    }

    public LocalDateTime getUpdatedAt() {
        return updatedAt;
    }

    public void setUpdatedAt(LocalDateTime updatedAt) {
        this.updatedAt = updatedAt;
    }

    public Integer getViewCount() {
        return viewCount;
    }

    public void setViewCount(Integer viewCount) {
        this.viewCount = viewCount;
    }

    public Boolean getFeatured() {
        return featured;
    }

    public void setFeatured(Boolean featured) {
        this.featured = featured;
    }

    public void incrementViewCount() {
        this.viewCount++;
    }
}

File: bo/types/article.ts
import { Member } from './member';

export enum ArticleCategory {
  NEWS = 'NEWS',
  RESULTS = 'RESULTS',
  EVENTS = 'EVENTS',
  TUTORIAL = 'TUTORIAL',
  ANNOUNCEMENT = 'ANNOUNCEMENT',
}

export enum ArticleStatus {
  DRAFT = 'DRAFT',
  PUBLISHED = 'PUBLISHED',
  ARCHIVED = 'ARCHIVED',
}

export interface Article {
  id: string; // UUID
  title: string;
  slug: string;
  content: string; // JSON content from Editor.js
  excerpt?: string;
  category: ArticleCategory;
  status: ArticleStatus;
  author: Member;
  coverImage?: string;
  tags?: string[];
  publishedAt?: string; // ISO date string
  createdAt: string; // ISO date string
  updatedAt: string; // ISO date string
  viewCount: number;
  featured: boolean;
}

export interface ArticleStats {
  total: number;
  published: number;
  draft: number;
  archived: number;
}

export interface ArticleFormData {
  title: string;
  slug?: string;
  content: string;
  excerpt?: string;
  category: ArticleCategory;
  coverImage?: string;
  tags?: string[];
  featured?: boolean;
}

export interface ArticleFilters {
  status?: ArticleStatus;
  category?: ArticleCategory;
  search?: string;
  page?: number;
  size?: number;
  sortBy?: string;
  sortDirection?: 'asc' | 'desc';
}

export interface ArticlePage {
  content: Article[];
  totalElements: number;
  totalPages: number;
  size: number;
  number: number;
  first: boolean;
  last: boolean;
}

// Labels for display
export const ArticleCategoryLabels: Record<ArticleCategory, string> = {
  [ArticleCategory.NEWS]: 'Actualités',
  [ArticleCategory.RESULTS]: 'Résultats',
  [ArticleCategory.EVENTS]: 'Événements',
  [ArticleCategory.TUTORIAL]: 'Tutoriels',
  [ArticleCategory.ANNOUNCEMENT]: 'Annonces',
};

export const ArticleStatusLabels: Record<ArticleStatus, string> = {
  [ArticleStatus.DRAFT]: 'Brouillon',
  [ArticleStatus.PUBLISHED]: 'Publié',
  [ArticleStatus.ARCHIVED]: 'Archivé',
};

// Colors for status tags
export const ArticleStatusColors: Record<ArticleStatus, string> = {
  [ArticleStatus.DRAFT]: 'default',
  [ArticleStatus.PUBLISHED]: 'success',
  [ArticleStatus.ARCHIVED]: 'warning',
};
